components/TextFields/src/ContainedInputView/private/MDCContainedInputView.h

```diff
@@ -0,0 +1,107 @@
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+#import "MDCContainedInputViewColorViewModel.h"
+#import "MDCContainedInputViewLabelAnimation.h"
+#import "MDCContainedInputViewLabelState.h"
+#import "MDCContainerStyleVerticalPositioningReference.h"
+#import "MDCTextControlLabelBehavior.h"
+#import "MDCTextControlState.h"
+
+static const CGFloat kMDCContainedInputViewDefaultAnimationDuration = (CGFloat)0.15;
+
+@protocol MDCContainedInputViewStyle;
+
+@protocol MDCContainedInputView <NSObject>
+
+/**
+ Dictates the @c MDCContainedInputViewStyle of the text field. Defaults to an instance of
+ MDCContainedInputViewStyleBase.
+ */
+@property(nonatomic, strong, nonnull) id<MDCContainedInputViewStyle> containerStyle;
+
+/**
+ Describes the current @c MDCtextControlState of the view.
+ */
+@property(nonatomic, assign, readonly) MDCTextControlState textControlState;
+
+/**
+ Describes the current @c MDCContainedInputViewLabelState of the contained input view. This
+ value is affected by things like the view's state, the value for @c canFloatingLabelFloat, and the
+ text of the floating label.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewLabelState labelState;
+
+/**
+ The @c label is a label that occupies the text area when there is no text and that floats
+ above the text once there is some. It is distinct from a placeholder.
+ */
+@property(strong, nonatomic, readonly, nonnull) UILabel *label;
+
+/**
+ The @c normalFont is the contained input view's primary font. The text is this font. The label also
+ has this font when it isn't floating.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIFont *normalFont;
+
+/**
+ The @c floatingFont is the font of the label when it's floating.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIFont *floatingFont;
+
+/**
+ This method returns a color scheme for a given state.
+ */
+- (nonnull MDCContainedInputViewColorViewModel *)containedInputViewColorViewModelForState:
+    (MDCTextControlState)textControlState;
+
+/**
+ This method sets a color scheme for a given state.
+ */
+- (void)setContainedInputViewColorViewModel:
+            (nonnull MDCContainedInputViewColorViewModel *)containedInputViewColorViewModel
+                                   forState:(MDCTextControlState)textFieldState;
+
+@end
+
+@protocol MDCContainedInputViewStyle <NSObject>
+
+/**
+ This method allows objects conforming to MDCContainedInputViewStyle to apply themselves to objects
+ conforming to MDCContainedInputView with a set of colors represented by an object conforming to
+ MDCContainedInputViewColorViewModel.
+ */
+- (void)applyStyleToContainedInputView:(nonnull id<MDCContainedInputView>)containedInputView;
+/**
+ This method allows objects conforming to MDCContainedInputViewStyle to remove the styling
+ previously applied to objects conforming to MDCContainedInputView.
+ */
+- (void)removeStyleFrom:(nonnull id<MDCContainedInputView>)containedInputView;
```

Comment:
why do we need to remove styles? isn't just setting a new style enough?

Reply:
Originally, `containerStyle` was going to be an exposed property. The idea was that clients were going to be allowed to do things like this:
```objc
textField.containerStyle = [MDCContainedInputViewStyleFilled new];
// it's a filled textfield now
textField.containerStyle = [MDCContainedInputViewStyleOutlined new];
// it's an outlined textfield now
```
That's no longer the case though. This could be removed and nothing bad would happen. For a better idea of how it might be implemented, here's what the MDCContainedInputViewStyleOutlined implementation looks like:
```objc
- (void)removeStyleFrom:(id<MDCContainedInputView>)containedInputView {
  [self.outlinedSublayer removeFromSuperlayer];
}
```